File: candle-flow/backend/app/core/windows.py
"""Window (gap) zones: Nison treats them as support/resistance *areas*.

A window is filled only when a later *close* re-enters the gap — wicks do not count.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Optional, Sequence

from app.core.indicators import is_downtrend, is_uptrend


@dataclass
class WindowZone:
    kind: str  # rising | falling
    start_index: int
    top: float
    bottom: float
    filled_index: Optional[int] = None

    @property
    def filled(self) -> bool:
        return self.filled_index is not None

    @property
    def key_edge(self) -> float:
        """Rising: lower rim (support). Falling: upper rim (resistance)."""
        return self.bottom if self.kind == "rising" else self.top


def collect_windows(klines: Sequence, min_gap_pct: float = 0.003) -> list[WindowZone]:
    """Find rising/falling windows; mark fill by later close."""
    if len(klines) < 10:
        return []
    zones: list[WindowZone] = []
    for i in range(1, len(klines)):
        prev, cur = klines[i - 1], klines[i]
        prev_high, prev_low = float(prev.high), float(prev.low)
        cur_high, cur_low = float(cur.high), float(cur.low)
        px = max(float(prev.close), 0.01)
        if cur_low > prev_high and (cur_low - prev_high) / px >= min_gap_pct and is_uptrend(klines, i):
            zones.append(WindowZone("rising", i, top=cur_low, bottom=prev_high))
        elif cur_high < prev_low and (prev_low - cur_high) / px >= min_gap_pct and is_downtrend(klines, i):
            zones.append(WindowZone("falling", i, top=prev_low, bottom=cur_high))

    for z in zones:
        for j in range(z.start_index + 1, len(klines)):
            close = float(klines[j].close)
            if z.kind == "rising" and close <= z.bottom:
                z.filled_index = j
                break
            if z.kind == "falling" and close >= z.top:
                z.filled_index = j
                break
    return zones


def active_windows(klines: Sequence, index: int) -> list[WindowZone]:
    """Windows still open as of `index` (fill on or before index excludes them)."""
    out = []
    for z in collect_windows(klines):
        if z.start_index >= index:
            continue
        if z.filled_index is not None and z.filled_index <= index:
            continue
        out.append(z)
    return out


def window_still_open(klines: Sequence, start_index: int, kind: str) -> bool:
    for z in collect_windows(klines):
        if z.start_index == start_index and z.kind == kind:
            return not z.filled
    return False

File: candle-flow/frontend/src/utils/windows.ts
import type { KlineItem } from '@/api'

export type WindowZone = {
  kind: 'rising' | 'falling'
  startIndex: number
  top: number
  bottom: number
  filledIndex: number | null
  filled: boolean
  keyEdge: number
}

function sma(closes: number[], period: number, end: number): number | null {
  if (end < period - 1) return null
  let sum = 0
  for (let i = end - period + 1; i <= end; i++) sum += closes[i]
  return sum / period
}

function inUptrend(closes: number[], index: number) {
  const ma5 = sma(closes, 5, index)
  const ma10 = sma(closes, 10, index)
  const ma20 = sma(closes, 20, index)
  return !!(ma5 && ma10 && ma20 && ma5 > ma10 && ma10 > ma20)
}

function inDowntrend(closes: number[], index: number) {
  const ma5 = sma(closes, 5, index)
  const ma10 = sma(closes, 10, index)
  const ma20 = sma(closes, 20, index)
  return !!(ma5 && ma10 && ma20 && ma5 < ma10 && ma10 < ma20)
}

/** Rising/falling windows; a window is filled only when a later *close* re-enters the gap. */
export function collectWindows(klines: KlineItem[], minGapPct = 0.003): WindowZone[] {
  if (klines.length < 22) return []
  const closes = klines.map((k) => Number(k.close))
  const zones: WindowZone[] = []
  for (let i = 1; i < klines.length; i++) {
    const prev = klines[i - 1]
    const cur = klines[i]
    const prevHigh = Number(prev.high)
    const prevLow = Number(prev.low)
    const curHigh = Number(cur.high)
    const curLow = Number(cur.low)
    const px = Math.max(Number(prev.close), 0.01)
    if (curLow > prevHigh && (curLow - prevHigh) / px >= minGapPct && inUptrend(closes, i)) {
      zones.push({
        kind: 'rising',
        startIndex: i,
        top: curLow,
        bottom: prevHigh,
        filledIndex: null,
        filled: false,
        keyEdge: prevHigh,
      })
    } else if (curHigh < prevLow && (prevLow - curHigh) / px >= minGapPct && inDowntrend(closes, i)) {
      zones.push({
        kind: 'falling',
        startIndex: i,
        top: prevLow,
        bottom: curHigh,
        filledIndex: null,
        filled: false,
        keyEdge: prevLow,
      })
    }
  }
  for (const z of zones) {
    for (let j = z.startIndex + 1; j < klines.length; j++) {
      const close = Number(klines[j].close)
      if (z.kind === 'rising' && close <= z.bottom) {
        z.filledIndex = j
        z.filled = true
        break
      }
      if (z.kind === 'falling' && close >= z.top) {
        z.filledIndex = j
        z.filled = true
        break
      }
    }
  }
  return zones
}

export function unfilledWindows(klines: KlineItem[], maxCount = 3): WindowZone[] {
  return collectWindows(klines).filter((z) => !z.filled).slice(-maxCount)
}
